### ISSUE TITLE
`AbstractSqlBuilder` generates incorrect `case` statement for boolean type

### ISSUE THREAD

**Issue by jameschenjav** [juanluispaz/ts-sql-query#90]
https://github.com/juanluispaz/ts-sql-query/blob/6794c0c179ec434e095b48a509a3089946dda51f/src/sqlBuilders/AbstractSqlBuilder.ts#L1194

tldr:
the case is Oracle DB with a custom boolean type
```ts
class Foo extends Table<DBConnection, 'Foo'> {
  // ...
  is_bar = this.optionalColumnWithDefaultValue('is_bar', 'boolean', new CustomBooleanTypeAdapter('Y', 'N'));
  // ...
}

const tFoo = new Foo();

// update
const testCase = async () => {
  const connection = await getOracleDbConnection();
  const updateStatement = connection
    .update(tFoo)
    .set({
      is_bar:  true,
    })
    .where(tFoo.id.equals(42));
  console.log({ query: updateStatement.query });
  await updateStatement.executeUpdate();
};
```

it will get an error
```text
error: #req-1 ORA-00920: invalid relational operator
 ×ChainedError 500: ORA-00920: invalid relational operator: ORA-00920: invalid relational operator
```

it generates statement like this
```sql
update foo set is_bar = case when (:0 = 1) then 'Y' when not :1 then 'N' else null end where id = :2

case
when (:0 = 1) then 'Y'
when not :1 then 'N'
else null
end
```

it should be
```sql
case
when (:0 = 1)
  then 'Y'
when not (:1 = 1)
  then 'N'
else
  null
end
```

current code is this
```ts
// remapped
'case ' + 
'when ' + this._appendConditionValue(value, params, columnType, columnTypeAdapter) +
  ' then ' + this._appendLiteralValue(columnTypeAdapter.trueValue, params) +
' when not ' + this._appendValue(value, params, columnType, columnTypeAdapter) + // _appendValue
  ' then ' + this._appendLiteralValue(columnTypeAdapter.falseValue, params) +
' else null end'
```

it should be
```ts
// remapped
'case ' + 
'when ' + this._appendConditionValue(value, params, columnType, columnTypeAdapter) +
  ' then ' + this._appendLiteralValue(columnTypeAdapter.trueValue, params) +
' when not ' + this._appendConditionValue(value, params, columnType, columnTypeAdapter) + // _appendConditionValue
  ' then ' + this._appendLiteralValue(columnTypeAdapter.falseValue, params) +
' else null end'
```

then I searched `when not` in the codebase, there are other 9 results, basically doing similar things, like
```ts
// src/sqlBuilders/SqlServerSqlBuilder.ts:48
return 'cast(case when ' + super._appendConditionSql(value, params) + ' then 1 when not ' + super._appendConditionSql(value, params) + ' then 0 else null end as bit)'

// src/sqlBuilders/OracleSqlBuilder.ts:46
return 'case when ' + super._appendConditionSql(value, params) + ' then 1 when not ' + super._appendConditionSql(value, params) + ' then 0 else null end'
```

**Comment by jameschenjav** [comment]
i also tried with
```ts
`case ${this._appendValue(value, params, columnType, columnTypeAdapter)} when 1 then ${this._appendLiteralValue(columnTypeAdapter.trueValue, params)} when 0 then ${this._appendLiteralValue(columnTypeAdapter.falseValue, params)} else null end`
```

it generates
```sql
update foo
set is_bar = case :0 when 1 then 'Y' when 0 then 'N' else null end
where id = :1
```

which makes more sense to me. it works in oracle, but i am not sure if this going to work for others

**Comment by juanluispaz** [comment]
Oh, small fail, Boolean values in Sql Server and Oracle are headache.

I like the second approach because the one binding, and it works for the other databases; I'm going to change it.

**Comment by juanluispaz** [comment]
There is something calling my attention: you defined the column as `optionalColumnWithDefaultValue`, that means, the column can be null; is that correct, it exists `columnWithDefaultValue` where the column doesn't admit null values in the table. An optional column with default value is a odd choice in tables; it is possible, but, in most common cases is indicator of a mistake

**Comment by jameschenjav** [comment]
> that means, the column can be null

We do have a lot of `alter table FOO add BAR VARCHAR2(1) default 'N'`. So it is nullable and with a default value. Weird choice, but it is what it is.

I will suggest changing it to `CHAR(1) NOT NULL default 'N'` with a check constraint on it. But I can't simply make the call. As you said, boolean type in oracle just likes a pain in the ass. I wish we had a better table design.

**Comment by jameschenjav** [comment]
btw, thx a lot for building this lib. i really appreciate your quick responses.

i have tried several ORMs/builders. Kysely looks great, but does not support oracle. i also tried knex, but seems have to use a separated pool, and all the identities must be upper-case, and quite confusing due to no difference between a column name and a normal string value.

the approach of `ts-sql-query` is sane to me. it's the best bet for oracle if we will move to node + ts back-end.

**Comment by juanluispaz** [comment]
Published ts-sql-query 1.50.0 with a major rework on custom booleans to fix several bugs; I added extensive tests with real databases (for all supported databases) to ensure everything is working as expected.

 ts-sql-query tries to use the most simple form of boolean transformation when possible, but some corner cases require a more complex form with more than one binding. It is possible some SQL can be optimised in specific cases I didn't identify at this moment.

Try and let me know your feedback.

> btw, thx a lot for building this lib. i really appreciate your quick responses.

You're welcome, and let me know any suggestions, requests or doubts. I will be pleased to help you. Also, be aware ts-sql-query already supports database-specific features validated during compile time; if you miss something, let me know.
